pkg/collector/udp.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2020 VMware, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"bytes"
+	"github.com/vmware/go-ipfix/pkg/config"
+	"net"
+	"sync"
+	"time"
+
+	"k8s.io/klog/v2"
+)
+
+func (cp *collectingProcess) startUDPServer() {
+	s, err := net.ResolveUDPAddr("udp", cp.address.String())
+	if err != nil {
+		klog.Error(err)
+		return
+	}
+	conn, err := net.ListenUDP("udp", s)
+	if err != nil {
+		klog.Error(err)
+		return
+	}
+	klog.Infof("Start %s collecting process on %s", cp.address.Network(), cp.address.String())
+	var wg sync.WaitGroup
+	defer conn.Close()
+	go func() {
+		for {
+			buff := make([]byte, cp.maxBufferSize)
+			size, address, err := conn.ReadFromUDP(buff)
+			cp.handleUDPClient(address, &wg)
+			cp.clients[address.String()].packetChan <- bytes.NewBuffer(buff[0:size])
+			if err != nil {
+				klog.Errorf("Error in collecting process: %v", err)
+				return
+			}
+		}
+	}()
+	select {
+	case <-cp.stopChan:
+		// stop all the workers before closing collector
+		for _, client := range cp.clients {
+			client.errChan <- true
+		}
+		wg.Wait()
+		return
+	}
+}
+
+func (cp *collectingProcess) handleUDPClient(address net.Addr, wg *sync.WaitGroup) {
+	if _, exist := cp.clients[address.String()]; !exist {
+		client := cp.createClient()
+		cp.clients[address.String()] = client
+		wg.Add(1)
+		defer wg.Done()
+		go func() {
+			ticker := time.NewTicker(time.Duration(config.TemplateRefreshTimeOut) * time.Second)
+			for {
+				select {
+				case <-client.errChan:
+					klog.Infof("UDP connection from %s has been closed.", address.String())
+					return
+				case <-ticker.C: // set timeout for udp connection
+					klog.Errorf("UDP connection from %s timed out.", address.String())
+					cp.deleteClient(address.Network())
+					return
+				case packet := <-client.packetChan:
+					// get the message here
+					message, err := cp.decodePacket(packet)
+					if err != nil {
+						klog.Error(err)
+						return
+					}
+					klog.Info(message)
```

Comment:
Increase the debug log level to 2 or 4.  Seeing all messages in the log is not desirable.